src/EditorFeatures/Core/Intents/IIntentProcessor.cs

```diff
@@ -0,0 +1,85 @@
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Features.Intents;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.VisualStudio.Text;
+
+namespace Microsoft.CodeAnalysis.Editor.Intents
+{
+    internal interface IIntentProcessor
+    {
+        /// <summary>
+        /// For an input intent, computes the edits required to apply that intent and returns them.
+        /// </summary>
+        /// <param name="intentRequestContext">the intents with the context in which the intent was found.</param>
+        /// <returns>the edits that should be applied to the current snapshot.</returns>
+        Task<IntentResult?> ComputeEditsAsync(IntentRequestContext intentRequestContext, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// Defines the data needed to compute the code action edits from an intent.
+    /// </summary>
+    internal struct IntentRequestContext
+    {
+        /// <summary>
+        /// The intent name.  <see cref="WellKnownIntents"/> contains all intents roslyn knows how to handle.
+        /// </summary>
+        public string IntentName { get; }
+
+        /// <summary>
+        /// JSON formatted data specific to the intent that must be deserialized into the appropriate object.
+        /// </summary>
+        public string? IntentData { get; }
+
+        /// <summary>
+        /// The text and selection in the document before changes were made triggering an intent.
+        /// <remarks>
+        /// For example, if the text typed was 'public Class(' which led to a generate ctor intent,
+        /// we expect a snapshot before any of the ctor was typed.
+        /// </remarks>
+        /// </summary>
+        public SnapshotSpan PriorSnapshotSpan { get; }
+
+        /// <summary>
+        /// The text snapshot and selection when <see cref="IIntentProcessor.ComputeEditsAsync"/>
+        /// was called to compute the text edits and against which the resulting text edits will be calculated.
+        /// </summary>
+        public SnapshotSpan CurrentSnapshotSpan { get; }
+
+        public IntentRequestContext(string intentName, SnapshotSpan priorSnapshotSpan, SnapshotSpan currentSnapshotSpan, string? intentData)
+        {
+            IntentName = intentName ?? throw new ArgumentNullException(nameof(intentName));
+            IntentData = intentData;
+            PriorSnapshotSpan = priorSnapshotSpan;
+            CurrentSnapshotSpan = currentSnapshotSpan;
+        }
+    }
+
+    /// <summary>
+    /// Defines the text changes needed to apply an intent.
+    /// </summary>
+    internal struct IntentResult
+    {
+        /// <summary>
+        /// The text changes that should be applied to the <see cref="IntentRequestContext.CurrentSnapshotSpan"/>
+        /// </summary>
+        public readonly ImmutableArray<TextChange> TextChanges;
+
+        /// <summary>
+        /// The title associated with this intent result.
+        /// </summary>
+        public readonly string Title;
```

Comment:
to intellicode

6.  We can return the refactoring title for the intent, would this be useful for you?

Reply:
6
The roslyn name of the refactoring is useful for displaying at different locations
Some meta data on the refactoring we can use to figure out what works or what does not work (if you have multiple result we would be able to see what works better for user)

Reply:
@pgroene 
Could you expand on what metadata would be useful?  I presume you'd like something to tell apart sub-actions across multiple invocations of the same intent.  Would just a unique identifier for each sub action be enough, or do you need to know something more?

@CyrusNajmabadi do we have anything like that?  The only thing I can think of is potentially the type of the code action that we give back

Reply:
>@CyrusNajmabadi do we have anything like that? The only thing I can think of is potentially the type of the code action that we give back

Not especially.  From our perspective, they should all work well.  We will already be ordering thrm from "most viable" to "least viable".

THat said, there are times when we're just up in the air as to what you want.  For example, if you do `this.x = 0;`, and you have no fields, we will offer to generate eitehr a normal field or a readonly field.

Reply:
@CyrusNajmabadi I think (and Peter can confirm) that they would like metadata to distinguish between the two actions so that from their telemetry they can capture which one the user prefers.  We can return them in the order we think is right, but then they could use telemetry to figure out the actual usage.  In that case I think type name may be appropriate?